File: Photino.API/Ns/IO.cs
using System;
using System.IO;
using System.Text;

namespace PhotinoAPI.Ns
{
    [PhotonName("io")]
    public class IO : PhotonApiBase
    {
        public IO(PhotonManager manager) : base(manager) { }
        
        [PhotonName("readFile")]
        public static string ReadFile(string path, string encoding = null) =>
            encoding == null ? File.ReadAllText(path) : File.ReadAllText(path, Encoding.GetEncoding(encoding));

        [PhotonName("listFiles")]
        public static string[] ListFiles(string path, string searchPattern = null, bool recursive = false)
        {
            if (searchPattern != null && recursive)
                return Directory.GetFiles(path, searchPattern, SearchOption.AllDirectories);
            if (searchPattern != null)
                return Directory.GetFiles(path, searchPattern);
            return recursive ? Directory.GetFiles(path, "*", SearchOption.AllDirectories) : Directory.GetFiles(path);
        }

        [PhotonName("listFolders")]
        public static string[] ListFolders(string path, string searchPattern = null, bool recursive = false)
        {
            if (searchPattern != null && recursive)
                return Directory.GetDirectories(path, searchPattern, SearchOption.AllDirectories);
            if (searchPattern != null)
                return Directory.GetDirectories(path, searchPattern);
            return recursive ? Directory.GetDirectories(path, "*", SearchOption.AllDirectories) : Directory.GetDirectories(path);
        }

        [PhotonName("writeFile")]
        public static void WriteFile(string path, string contents, string encoding = null)
        {
            if (encoding == null)
                File.WriteAllText(path, contents);
            else
                File.WriteAllText(path, contents, Encoding.GetEncoding(encoding));
        }

        [PhotonName("createFolder")]
        public static void CreateFolder(string path) => Directory.CreateDirectory(path);

        [PhotonName("deleteFile")]
        public static void DeleteFile(string path) => File.Delete(path);

        [PhotonName("deleteFolder")]
        public static void DeleteFolder(string path, bool recursive = false) => Directory.Delete(path, recursive);
        
        [PhotonName("fileExists")]
        public static bool FileExists(string path) => File.Exists(path);

        [PhotonName("cwd")]
        public static string CurrentFolder() => Environment.CurrentDirectory;
    }
}

File: TestApp/GUI/src/lib/photino/api/IO.ts
import type { Photino } from '..';
import APIBase from './APIBase';

export default class IO extends APIBase {
  constructor(photino: Photino) {
    super(photino);
  }

  readFile(path: string, encoding: BufferEncoding | null = null): Promise<string> {
    return this.photino.send({ ns: 'io', action: 'readFile', params: { path, encoding } });
  }

  writeFile(path: string, contents: string, encoding: BufferEncoding | null = null): Promise<void> {
    return this.photino.send({ ns: 'io', action: 'writeFile', params: { path, contents, encoding } });
  }

  listFiles(path: string, searchPattern: string = null, recursive: boolean = false): Promise<string[]> {
    return this.photino.send({ ns: 'io', action: 'listFiles', params: { path, searchPattern, recursive } });
  }

  listFolders(path: string, searchPattern: string = null, recursive: boolean = false): Promise<string[]> {
    return this.photino.send({ ns: 'io', action: 'listFolders', params: { path, searchPattern, recursive } });
  }

  createFolder(path: string): Promise<void> {
    return this.photino.send({ ns: 'io', action: 'createFolder', params: { path } });
  }

  deleteFile(path: string): Promise<void> {
    return this.photino.send({ ns: 'io', action: 'deleteFile', params: { path } });
  }

  deleteFolder(path: string, recursive: boolean = false): Promise<void> {
    return this.photino.send({ ns: 'io', action: 'deleteFolder', params: { path, recursive } });
  }

  cwd(): Promise<string> {
    return this.photino.send({ ns: 'io', action: 'cwd', params: {} });
  }
}
